server/src/main/java/org/elasticsearch/search/lookup/SourceLookup.java

```diff
@@ -204,72 +94,256 @@ public List<Object> extractRawValuesWithoutCaching(String path) {
+                return XContentMapValues.extractRawValues(path, XContentHelper.convertToMap(source, false, null, Set.of(path), null).v2());
+            } catch (Exception e) {
+                throw new ElasticsearchParseException("failed to parse / load source", e);
+            }
+        }
+
+        /**
+         * For the provided path, return its value in the source.
+         *
+         * Both array and object values can be returned.
+         *
+         * @param path the value's path in the source.
+         * @param nullValue a value to return if the path exists, but the value is 'null'. This helps
+         *                  in distinguishing between a path that doesn't exist vs. a value of 'null'.
+         *
+         * @return the value associated with the path in the source or 'null' if the path does not exist.
+         */
+        public Object extractValue(String path, @Nullable Object nullValue) {
+            return XContentMapValues.extractValue(path, source(), nullValue);
+        }
+
+        public Object filter(FetchSourceContext context) {
+            return context.getFilter().apply(source());
+        }
+
+        @Override
+        public Object get(Object key) {
+            return source().get(key);
+        }
+
+        @Override
+        public int size() {
+            return source().size();
+        }
+
+        @Override
+        public boolean isEmpty() {
+            return source().isEmpty();
+        }
+
+        @Override
+        public boolean containsKey(Object key) {
+            return source().containsKey(key);
+        }
+
+        @Override
+        public boolean containsValue(Object value) {
+            return source().containsValue(value);
+        }
+
+        @Override
+        public Set<String> keySet() {
+            return source().keySet();
+        }
+
+        @Override
+        public Collection<Object> values() {
+            return source().values();
+        }
+
+        @Override
+        public Set<Map.Entry<String, Object>> entrySet() {
+            return source().entrySet();
+        }
 
-    @Override
-    public Object remove(Object key) {
-        throw new UnsupportedOperationException();
+        @Override
+        public Object put(String key, Object value) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Object remove(Object key) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        @SuppressWarnings("rawtypes")
+        public void putAll(Map m) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void clear() {
+            throw new UnsupportedOperationException();
+        }
     }
 
-    @Override
-    @SuppressWarnings("rawtypes")
-    public void putAll(Map m) {
-        throw new UnsupportedOperationException();
+    class Static extends AbstractSourceLookup {
```

Comment:
These both deserve javadoc please.

I wonder if `PreLoaded` is a better name than `Static`. My mind jumps to java's `static` keyword and I keep thinking "there is only one of this". But static here means something more like "unchanging". 

Speaking of, how hard would be be to make the `Static` version truly immutable? Not worth it in this PR, I'd wager, but maybe actually good?